### PR TITLE
Add api to buffer donation

### PR DESCRIPTION
**Pull request by JackCaoG**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def get_git_head_sha(base_dir):
 
 
 def get_build_version(xla_git_sha):
-  version = os.getenv('TORCH_XLA_VERSION', '2.2.0')
+  version = os.getenv('TORCH_XLA_VERSION', '2.3.0')
```
Comment (will-cromar) [inline setup.py:91]: nit: this is correct, but please separate it into a new PR.

File: test/dynamo/test_dynamo.py [hunk 2]
```diff
@@ -286,7 +286,7 @@ def fn_fallback(t):
     xla_dynamo_res = dynamo_fn(t_xla)
     self.assertTrue(torch.allclose(cpu_res, xla_dynamo_res.cpu()))
     self.assertEqual(met.metric_data('CompileTime')[0], 3)
-    self.assertEqual(met.metric_data('ExecuteTime')[0], 9)
+    self.assertEqual(met.metric_data('ExecuteTime')[0], 7)
```
Comment (will-cromar) [inline test/dynamo/test_dynamo.py:289]: what does `7` represent?

File: torch_xla/core/dynamo_bridge.py [hunk 3]
```diff
@@ -28,6 +28,25 @@
 ptxla_debug = int(os.environ.get('PT_XLA_DEBUG', '0')) == 1
 
 
+class AliasWithBufferDonorContext(object):
+
+  def __init__(self, should_alias: bool):
```
Comment (will-cromar) [inline torch_xla/core/dynamo_bridge.py:33]: Can you make `should_alias` default to `True`? I think that's going to be the case most of the time when you actually invoke this context manager.

File: torch_xla/core/dynamo_bridge.py [hunk 4]
```diff
@@ -28,6 +28,25 @@
...
+    self.should_alias = should_alias
+
+  def __enter__(self):
```
Comment (will-cromar) [inline torch_xla/core/dynamo_bridge.py:36]: I strongly recommend using `contextlib.contextmanager` for trivial contexts like this: https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

File: torch_xla/core/dynamo_bridge.py [hunk 5]
```diff
@@ -28,6 +28,25 @@
...
+    self.env_inited = 'XLA_SHOULD_ALIAS_WITH_BUFFER_DONOR' in os.environ
+    if self.env_inited:
+      self.env_saved = os.environ['XLA_SHOULD_ALIAS_WITH_BUFFER_DONOR']
+    os.environ[
```
Comment (will-cromar) [inline torch_xla/core/dynamo_bridge.py:40]: We really need to break that habit of communicating between Python/C++ with env vars. Setting a variable in C++ doesn't update Python's `os.environ` for example.

Can you add a pybind to update an internal variable instead?

**Comment by JackCaoG** [comment]
oh.. I need to rebase this pr haha. My old pr for aliasing in https://github.com/pytorch/xla/pull/6587. I will adjust your comment there.

**Pull request closed without merging**